src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -1734,99 +1734,139 @@ private BoundPattern BindBinaryPattern(
             bool hasErrors,
             BindingDiagnosticBag diagnostics)
         {
-            bool isDisjunction = node.Kind() == SyntaxKind.OrPattern;
-            if (isDisjunction)
-            {
-                MessageID.IDS_FeatureOrPattern.CheckFeatureAvailability(diagnostics, node.OperatorToken);
+            // Users (such as ourselves) can have many, many nested binary patterns. To avoid crashing, do left recursion manually.
+
+            var binaryPatternStack = ArrayBuilder<(BinaryPatternSyntax pat, bool permitDesignations)>.GetInstance();
+            BinaryPatternSyntax? currentNode = node;
 
-                permitDesignations = false; // prevent designators under 'or'
-                var left = BindPattern(node.Left, inputType, permitDesignations, hasErrors, diagnostics);
-                var right = BindPattern(node.Right, inputType, permitDesignations, hasErrors, diagnostics);
+            do
+            {
+                permitDesignations = permitDesignations && currentNode.IsKind(SyntaxKind.AndPattern);
+                binaryPatternStack.Push((currentNode, permitDesignations));
+                currentNode = currentNode.Left as BinaryPatternSyntax;
+            } while (currentNode != null);
 
-                // Compute the common type. This algorithm is quadratic, but disjunctive patterns are unlikely to be huge
-                var narrowedTypeCandidates = ArrayBuilder<TypeSymbol>.GetInstance(2);
-                collectCandidates(left, narrowedTypeCandidates);
-                collectCandidates(right, narrowedTypeCandidates);
-                var narrowedType = leastSpecificType(node, narrowedTypeCandidates, diagnostics) ?? inputType;
-                narrowedTypeCandidates.Free();
+            Debug.Assert(binaryPatternStack.Count > 0);
 
-                return new BoundBinaryPattern(node, disjunction: isDisjunction, left, right, inputType: inputType, narrowedType: narrowedType, hasErrors);
+            BoundPattern? result = null;
 
-                static void collectCandidates(BoundPattern pat, ArrayBuilder<TypeSymbol> candidates)
+            while (binaryPatternStack.TryPop(out var binaryPatternAndPermitDesignations))
+            {
+                (var binaryPattern, permitDesignations) = binaryPatternAndPermitDesignations;
+                if (result == null)
                 {
-                    if (pat is BoundBinaryPattern { Disjunction: true } p)
-                    {
-                        collectCandidates(p.Left, candidates);
-                        collectCandidates(p.Right, candidates);
-                    }
-                    else
-                    {
-                        candidates.Add(pat.NarrowedType);
-                    }
+                    Debug.Assert(binaryPattern.Left is not BinaryPatternSyntax);
+                    result = BindPattern(binaryPattern.Left, inputType, permitDesignations, hasErrors, diagnostics);
```

Comment:
>result = BindPattern(binaryPattern.Left, inputType, permitDesignations, hasErrors, diagnostics); [](http://example.com/codeflow?start=20&length=96)

I think the logic could be simpler and much easier to follow if this conditional `result` initialization was done unconditionally before the loop. The node to bind could be "picked" from the stack, or could be stored in a local by the previous loop.